libportal/inputcapture.c

```diff
@@ -0,0 +1,1207 @@
+                  0,
+                  NULL, NULL,
+                  NULL,
+                  G_TYPE_NONE, 2,
+                  G_TYPE_UINT,
+                  G_TYPE_VARIANT);
+
+  /**
+   * XdpInputCaptureSession::disabled:
+   * @session: the [class@InputCaptureSession]
+   * @options: a GVariant with the signal options
+   *
+   * Emitted when an InputCapture session is disabled. This signal
+   * is emitted when capturing was disabled by the server.
+   */
+  signals[SIGNAL_DISABLED] =
+    g_signal_new ("disabled",
+                  G_TYPE_FROM_CLASS (object_class),
+                  G_SIGNAL_RUN_CLEANUP | G_SIGNAL_NO_RECURSE | G_SIGNAL_NO_HOOKS,
+                  0,
+                  NULL, NULL,
+                  NULL,
+                  G_TYPE_NONE, 1,
+                  G_TYPE_VARIANT);
+}
+
+static void
+xdp_input_capture_session_init (XdpInputCaptureSession *session)
+{
+  session->parent_session = NULL;
+  session->zones = NULL;
+  session->zone_set = 0;
+  for (guint i = 0; i < SIGNAL_LAST_SIGNAL; i++)
+    session->signal_ids[i] = 0;
+}
+
+/* A request-based method call */
+typedef struct {
+  XdpPortal *portal;
+  char *session_path; /* object path for session */
+  GTask *task;
+  guint signal_id; /* Request::Response signal */
+  char *request_path; /* object path for request */
+  guint cancelled_id; /* signal id for cancelled gobject signal */
+
+  /* CreateSession only */
+  XdpParent *parent;
+  char *parent_handle;
+  XdpCapability capabilities;
+
+  /* GetZones only */
+  XdpInputCaptureSession *session;
+
+  /* SetPointerBarrier only */
+  GList *barriers;
+
+} Call;
+
+static void create_session (Call *call);
+static void get_zones (Call *call);
+
+static void
+call_free (Call *call)
+{
+  /* CreateSesssion */
+  if (call->parent)
+    {
+      call->parent->parent_unexport (call->parent);
+      xdp_parent_free (call->parent);
+    }
+  g_free (call->parent_handle);
+
+  /* Generic */
+  if (call->signal_id)
+    g_dbus_connection_signal_unsubscribe (call->portal->bus, call->signal_id);
+
+  if (call->cancelled_id)
+    g_signal_handler_disconnect (g_task_get_cancellable (call->task), call->cancelled_id);
+
+  g_free (call->request_path);
+
+  if (call->portal)
+    g_object_unref (call->portal);
+  if (call->task)
+    g_object_unref (call->task);
+  if (call->session)
+    g_object_unref (call->session);
+
+  g_free (call->session_path);
+
+  g_free (call);
+}
+
+static void
+call_returned (GObject *object,
+               GAsyncResult *result,
+               gpointer data)
+{
+  Call *call = data;
+  GError *error = NULL;
+  g_autoptr(GVariant) ret;
+
+  ret = g_dbus_connection_call_finish (G_DBUS_CONNECTION (object), result, &error);
+  if (error)
+    {
+      if (call->cancelled_id)
+        {
+          g_signal_handler_disconnect (g_task_get_cancellable (call->task), call->cancelled_id);
+          call->cancelled_id = 0;
+        }
```

Comment:
```suggestion
      g_clear_signal_handler (&call->cancelled_id, g_task_get_cancellable (call->task));
```

Reply:
Done.

Reply:
Undone, sorry, that breaks the build (CI) with Ubuntu 20.04.